### PR TITLE
 Ignore static Main methods in NamingStyle analyzer

### LINKED ISSUE #49648: IDE1006 shouldn't report on top-level `Main` method with `await` operators
**Version Used**: 
VS 16.8.2

**Steps to Reproduce**:

1. Write a program using top-level statements, such as:
```csharp
// Program.cs
using System;

var sessionSettings = new SessionSettings("127.0.0.1", 8080, "1234567890");
await using var session = new Session(sessionSettings, myNumber);
await session.ConnectAsync();

Console.WriteLine("Successful!");
```

2. IDE1006 (naming styles) suggests.

**Expected Behavior**:
Because of implicit `Main` statement, IDE1006 shouldn't report on this scenario. In fact, we can't modify the suffix of the name of the `Main` method explicitly.

**Actual Behavior**:
IDE1006 raised, and all codes are marked with green waves.

### LINKED ISSUE #50734: Naming rule analyzer should not suggest MainAsync
**Version Used**: 

`Microsoft.CodeAnalysis.CSharp.NetAnalyzers` 5.0.102
`Microsoft.CodeAnalysis.NetAnalyzers` 5.0.102

**Steps to Reproduce**:

1. Create a new Console project
2. Setup `dotnet_naming_rule.async_methods_should_be_suffixed` in an .editorconfig file.
3. Write a `static async Task Main()` method.

**Expected Behavior**:

Since `MainAsync()` is not a valid application entry point, this analyzer rule should manually exclude valid `Main()` methods.

**Actual Behavior**:

The analyzer produces "IDE1006: Naming rule violation: Missing suffix: 'Async'" for `async` `Main()` methods.

### PR DESCRIPTION
**Pull request by Youssef1313**
Closes #49648
Fixes #50734

### CODE REVIEW

File: src/Analyzers/CSharp/Tests/NamingStyles/NamingStylesTests.cs [hunk 1]
```diff
@@ -1382,5 +1382,28 @@ class C : I
...
 }", new TestParameters(options: s_options.PropertyNamesArePascalCase));
         }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.NamingStyle)]
+        public async Task TestAsyncEntryPoint()
+        {
+            await TestMissingInRegularAndScriptAsync(@"
+class C
+{
+    async void [|Main|]()
```
Comment (Youssef1313) [inline src/Analyzers/CSharp/Tests/NamingStyles/NamingStylesTests.cs:1392]:
```suggestion
    static async void [|Main|]()
```

File: src/Analyzers/CSharp/Tests/NamingStyles/NamingStylesTests.cs [hunk 2]
```diff
@@ -1382,5 +1382,28 @@ class C : I
...
+
+        [Fact, Trait(Traits.Feature, Traits.Features.NamingStyle)]
+        public async Task TestAsyncEntryPoint()
+        {
+            await TestMissingInRegularAndScriptAsync(@"
+class C
+{
+    static async void [|Main|]()
+    {
+    }
```
Comment (Youssef1313) [inline src/Analyzers/CSharp/Tests/NamingStyles/NamingStylesTests.cs:1394]:
```suggestion
using System.Threading.Tasks;

class C
{
    static async Task Main()
    {
        await Task.Delay(0);
    }
```

File: src/Analyzers/CSharp/Tests/NamingStyles/NamingStylesTests.cs [hunk 3]
```diff
@@ -1382,5 +1382,31 @@ class C : I
...
+
+        [Fact, Trait(Traits.Feature, Traits.Features.NamingStyle)]
+        public async Task TestAsyncEntryPoint()
+        {
+            await TestMissingInRegularAndScriptAsync(@"
+using System.Threading.Tasks;
+
+class C
+{
+    static async Task Main()
```
Comment (Youssef1313) [inline src/Analyzers/CSharp/Tests/NamingStyles/NamingStylesTests.cs:1394]:
```suggestion
using System.Threading.Tasks;

class C
{
    static async Task [|Main|]()
```

File: src/Analyzers/CSharp/Tests/NamingStyles/NamingStylesTests.cs [hunk 4]
```diff
@@ -1382,5 +1382,31 @@ class C : I
...
+
+        [Fact, Trait(Traits.Feature, Traits.Features.NamingStyle)]
+        public async Task TestAsyncEntryPoint()
+        {
+            await TestMissingInRegularAndScriptAsync(@"
+using System.Threading.Tasks;
+
+class C
+{
+    static async Task [|Main|]()
```
Comment (Youssef1313) [inline src/Analyzers/CSharp/Tests/NamingStyles/NamingStylesTests.cs:1394]: This is different than the linked issue, but it fails currently. I'll work on getting that fix.

File: src/Analyzers/CSharp/Tests/NamingStyles/NamingStylesTests.cs [hunk 5]
```diff
@@ -1382,5 +1382,31 @@ class C : I
     int [|global::I.X|] => 0;
 }", new TestParameters(options: s_options.PropertyNamesArePascalCase));
         }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.NamingStyle)]
```
Comment (Youssef1313) [inline src/Analyzers/CSharp/Tests/NamingStyles/NamingStylesTests.cs:1386]:
```suggestion
        [Fact, Trait(Traits.Feature, Traits.Features.NamingStyle)]
        [WorkItem(50734, "https://github.com/dotnet/roslyn/issues/50734")]
```

**Comment by Youssef1313** [comment]
@CyrusNajmabadi Can you review please? Thanks.

File: src/Analyzers/Core/Analyzers/NamingStyle/NamingStyleDiagnosticAnalyzerBase.cs [hunk 6]
```diff
@@ -108,7 +108,9 @@ void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
             ConcurrentDictionary<Guid, ConcurrentDictionary<string, string?>> idToCachedResult,
             CancellationToken cancellationToken)
         {
-            if (string.IsNullOrEmpty(symbol.Name))
+            if (string.IsNullOrEmpty(symbol.Name) ||
+                // Heuristic for recognizing entry points.
+                symbol.IsStatic && symbol.Name == WellKnownMemberNames.EntryPointMethodName)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/NamingStyle/NamingStyleDiagnosticAnalyzerBase.cs:113]: we have an existing helper for this: IsEntryPoint

can you extrac tthis out (an extension on IMethodSymbol would be fine) and use the same for both?

Comment (jmarolf) [reply src/Analyzers/Core/Analyzers/NamingStyle/NamingStyleDiagnosticAnalyzerBase.cs:113]: yeah, [this method](https://github.com/dotnet/roslyn/blob/main/src/Analyzers/Core/Analyzers/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs#L702-L708) should be extracted out to a general helper and used here.

**Review by jmarolf — APPROVED** [review]
:shipit:

**Comment by jmarolf** [comment]
thanks @Youssef1313 !

**Pull request merged**